### DISCUSSION TITLE
How to Safely Use Kombu in a Multithreaded Environment?

### DISCUSSION THREAD

**Discussion by worl2160** [celery/kombu#2193] [Q&A]
I’m working on a project where I need to use the Kombu library for message passing, but I’ve encountered issues when running it in a multithreaded environment. After some research, I found out that Kombu is not thread-safe by design, which might explain the problems I’m facing, such as:

- Unexpected behavior when multiple threads try to publish messages simultaneously.
- Occasional crashes when accessing Kombu connections from different threads.

I understand that Kombu uses AMQP connections and channels, which are not thread-safe. However, I need to use threading in my application to achieve concurrent processing.

Could someone provide guidance or best practices on how to handle Kombu in a multithreaded context?

**Comment by TayfurYldz** [comment]
The safe unit to share between threads is the *pool*, not an acquired Kombu connection/channel/producer.

Kombu's current pool implementation is built around a thread-safe `LifoQueue`, and the documentation explicitly notes that global connection limits are normally process-wide while custom pool groups are available if you need per-thread pools. An acquired resource, however, has exclusive ownership until it is released; `Resource.acquire()` even documents that after release the caller must discard that object and acquire a new one before using it again.

For publishers, a straightforward pattern is:

```python
from kombu import Connection
from kombu.pools import producers

connection = Connection(BROKER_URL)

def publish_from_worker(payload):
    with producers[connection].acquire(block=True, timeout=10) as producer:
        producer.publish(payload, exchange=exchange, routing_key=key)
```

Each worker borrows a producer (and therefore its underlying connection/channel) for the duration of that operation, then returns it. Do not create one `Producer` or one `connection.default_channel` in the main thread and call it concurrently from all workers.

The same idea applies if you need lower-level access: use `connections[connection].acquire(...)`, and create/use the channel while that connection is owned by that worker. Set a finite pool limit appropriate for the broker and use `block=True` plus a timeout so concurrency is bounded rather than creating unbounded broker connections.

For consumers, I would go further and keep each consumer's connection/channel confined to the thread running its event/drain loop. Cross-thread coordination should happen through a thread-safe application queue/event, not by having another thread manipulate the consumer's channel.

Also avoid carrying these live resources across a process fork; Kombu's connection pools explicitly register cleanup after fork because inherited sockets are not safe to keep using as if they belonged to the child.
